src/mscorlib/src/System/Collections/Generic/Comparer.cs

```diff
@@ -77,9 +108,26 @@ private static Comparer<T> CreateComparer()
             return 0;
         }
     }
-
+    
+    // Helper class: Represents Comparer subclasses that do not have any fields.
+    // Their identity can be calculated as a function of their type.
     [Serializable]
-    internal class GenericComparer<T> : Comparer<T> where T: IComparable<T>
+    internal abstract class StatelessComparer<T> : Comparer<T>
```

Comment:
I do not think that the extra level of inheritance is a win here. The generic instantiations over value types are not small (couple hundred bytes each). This extra level of inheritance will make us create two instead of just one.

This should inherit directly like the `EqualityComparer`.